### DISCUSSION TITLE
how to use Picture Parallel Encoding in VVenC

### DISCUSSION THREAD

**Discussion by lingxiao6018855** [fraunhoferhhi/vvenc#275] [Q&A] [answered]
Hi !
I am currently conducting research on the coding quality and coding speed of vvenc encoding using different parallelization strategies/combinations. While reading your paper(Efficient Multi-Threading Strategies in VVenC,an Open and Optimized VVC Encoder Implementation) and considering the information provided in the project, I have not come across any specific content related to Picture Parallel Encoding cause I wanna do the test combine CTU and  Pic 。 I wanna know how to use that ,just like the usage of the WPP (Wavefront Parallel Processing) encoding strategy using "--WaveFrontSynchro=1". I would greatly appreciate it if you could provide me with information on how to use Picture Parallel Encoding in vvenc and any relevant details about Picture Parallel Encoding. 
Thank you very much in advance for your assistance.

best regard 
Xiao

**Comment by adamjw24** [comment, marked as answer]
The CTU encoding strategy is always enabled when more than 1 threads are used. It cannot be disabled without code modification.

The picture parallel processing is enabled per default, and can be disabled by setting `MaxParallelFrames=1`. Otherwise it is capped to either number of used threads, or 2 or 4 for rate-control and fixed-qp, respectively.

Hope this helps.

**Reply by lingxiao6018855** [reply]
thanks for your reply which really help me a lot !
best wishes 
Xiao

**Reply by lingxiao6018855** [reply]
Hi !
I'm sorry, I would like to ask you a question: After comparing the latest version of VTM and vvenc, I found that VTM seems to lack NEON optimization for the ARM architecture. This is because according to the following code:

_#if defined(__i386__) || defined(i386) || defined(__x86_64__) || defined(_M_X64) || defined (_WIN32) || defined (_MSC_VER)
#define TARGET_SIMD_X86
#elif defined (__ARM_NEON__)
#define TARGET_SIMD_ARM 1
#else
#error no simd target
#endif_


I couldn't find any relevant usage of TARGET_SIMD_ARM, but only TARGET_SIMD_X86. However, vvenc includes it. Can you please clarify if this is one of the improvements of vvenc compared to VTM, or if my analysis of VTM is not comprehensive enough?

best regards 
Xiao

**Reply by adamjw24** [reply]
Hi. You are right. VTM does not have proper support for ARM, although it should compile (but without any SIMD opt.).

Best, Adam

**Reply by lingxiao6018855** [reply]
thanks for your help !

**Reply by lingxiao6018855** [reply]
Hello ! 
Im sorry  for asking so many questions, but one thing I'm curious about is why the vvenc library includes the simde library, which contains the NEON instruction set. However, when I studied the transform & quantization in vvenc, I didn't seem to find any SIMD optimizations for the ARM architecture, only SIMD optimizations for the x86 architecture. Did I miss something? However, when I run the vvencFFapp on macOS M1, the terminal shows SIMD=NEON. So, has vvenc been optimized for the ARM architecture, and could you provide me with some detailed information about vvenc's usage of NEON? Thank you very much!

Best regards ,
Xiao

**Reply by adamjw24** [reply]
The intrinsics are written for x86. For other architectures we use SIMDe, as a compile time translation layer. This way, the executable contains NEON machine code, even if compiled from x86 intrinsics syntax. Thats why the output says NEON, since this is in the end, what is being executed. We are looking at improving some of the code explicitly for NEON by writing dedicated code, but its a more long term goal.

**Reply by lingxiao6018855** [reply]
thanks for your reply !
